ENH Add Multiclass Brier Score Loss (#22046)
scikit-learn/scikit-learn: doc/modules/model_evaluation.rst

discrete predictions.

For binary classification with a true label :math:`y \in \{0,1\}`
and a probability estimate :math:`\hat{p} \approx \operatorname{Pr}(y = 1)`,
the log loss per sample is the negative log-likelihood
of the classifier given the true label:

.. math::

    L_{\log}(y, \hat{p}) = -\log \operatorname{Pr}(y|\hat{p}) = -(y \log (\hat{p}) + (1 - y) \log (1 - \hat{p}))

This extends to the multiclass case as follows.
Let the true labels for a set of samples
be encoded as a 1-of-K binary indicator matrix :math:`Y`,
i.e., :math:`y_{i,k} = 1` if sample :math:`i` has label :math:`k`
taken from a set of :math:`K` labels.
Let :math:`\hat{P}` be a matrix of probability estimates,

> with :math:`\hat{p}_{i,k} \approx \operatorname{Pr}(y_{i,k} = 1)`.

Reviewer: ```suggestion
with elements :math:`\hat{p}_{i,k} \approx \operatorname{Pr}(y_{i,k} = 1)`.
```

Brier score loss
----------------

The :func:`brier_score_loss` function computes the `Brier score
<https://en.wikipedia.org/wiki/Brier_score>`_ for binary or multiclass
probabilistic predictions. Quoting Wikipedia:

    "The Brier score is a proper score function that measures the accuracy of
    probabilistic predictions. It is applicable to tasks in which predictions

>     must assign probabilities to a set of mutually exclusive discrete outcomes."

Reviewer 2: I don't understand the suggestion, what is the purpose or meaning of `[...] [It]` ?

> Let the true labels for a set of :math:`N` samples be encoded as a 1-of-K binary

Author: We decided to try refraining from using "samples" when referring to "data points"/"observations"/"records" in our English doc as a "sample" could represent a full dataset (e.g. the training set) and the reader could get confused as a result.

```suggestion
Let the true labels for a set of :math:`N` data points be encoded as a 1-of-K binary
```

Note that we don't plant to rename `n_samples` to `n_data_points` for the sake of backward compatibilities.

Brier score loss
----------------

The :func:`brier_score_loss` function computes the `Brier score
<https://en.wikipedia.org/wiki/Brier_score>`_ for binary and multiclass
probabilistic predictions and is equivalent to the mean squared error.
Quoting Wikipedia:

    "The Brier score is a strictly proper scoring rule that measures the accuracy of
    probabilistic predictions. [...] [It] is applicable to tasks in which predictions
    must assign probabilities to a set of mutually exclusive discrete outcomes or
    classes."

Let the true labels for a set of :math:`N` data points be encoded as a 1-of-K binary
indicator matrix :math:`Y`, i.e., :math:`y_{i,k} = 1` if sample :math:`i` has
label :math:`k` taken from a set of :math:`K` labels. Let :math:`\hat{P}` be a matrix

> of probability estimates :math:`\hat{p}_{i,k} \approx \operatorname{Pr}(y_{i,k} = 1)`.

Reviewer: ```suggestion
of probability estimates with elements :math:`\hat{p}_{i,k} \approx \operatorname{Pr}(y_{i,k} = 1)`.
```

achieves the best score only when the estimated probabilities are equal to the
true ones.

Note that in the binary case, the Brier score is usually divided by two and
ranges between :math:`[0,1]`. For binary targets :math:`y_i \in {0, 1}` and
probability estimates :math:`p_i  = \operatorname{Pr}(y_i = 1)` for the positive class,
the Brier score is then equal to:

.. math::

   BS(y, p) = \frac{1}{N} \sum_{i=0}^{N - 1}(y_i - p_i)^2

The :func:`brier_score_loss` function computes the Brier score given the
ground-truth labels and predicted probabilities,
as returned by an estimator's ``predict_proba`` method.

    >>> import numpy as np
    >>> from sklearn.metrics import brier_score_loss
    >>> y_true = np.array([0, 1, 1, 0])
    >>> y_true_categorical = np.array(["spam", "ham", "ham", "spam"])
    >>> y_prob = np.array([0.1, 0.9, 0.8, 0.4])
    >>> brier_score_loss(y_true, y_prob)
    0.055
    >>> brier_score_loss(y_true, 1 - y_prob, pos_label=0)
    0.055
    >>> brier_score_loss(y_true_categorical, y_prob, pos_label="ham")
    0.055
    >>> brier_score_loss(
    ...    ["eggs", "ham", "spam"],
    ...    [[0.8, 0.1, 0.1], [0.2, 0.7, 0.1], [0.2, 0.2, 0.6]],

>     ... )

Author: Our usage examples should always encourage users to pass `labels=` explicitly whenever the ordering is ambiguous (e.g. for non-integer classes).

Reviewer 2: done in a764d1f6a
